core/lib/mempool/src/mempool_store.rs

```diff
@@ -221,22 +221,53 @@ impl MempoolStore {
     }
 
     fn gc(&mut self) -> Vec<Address> {
-        if self.size >= self.capacity {
-            let index: HashSet<_> = self
+        if self.size > self.capacity {
+            let mut transactions = std::mem::take(&mut self.l2_transactions_per_account);
+            let mut possibly_kept: Vec<_> = self
                 .l2_priority_queue
                 .iter()
-                .map(|pointer| pointer.account)
+                .rev()
+                .filter_map(|pointer| {
+                    transactions
+                        .remove(&pointer.account)
+                        .map(|txs| (pointer.account, txs))
+                })
                 .collect();
-            let transactions = std::mem::take(&mut self.l2_transactions_per_account);
-            let (kept, drained) = transactions
+            let mut number_of_accounts_kept = possibly_kept
+                .iter()
+                .scan(0, |sum, (_, txs)| {
+                    *sum += txs.len();
+                    (*sum <= self.capacity as usize).then_some(())
+                })
+                .count();
+            if number_of_accounts_kept == 0 {
+                tracing::warn!("mempool capacity is too low to handle txs from single account, consider increasing capacity");
+                // Keep at least one entry, otherwise mempool won't return any new L2 tx to process.
+                number_of_accounts_kept = 1;
+            }
+            let (kept, drained) = {
+                let mut drained: Vec<_> = transactions.into_keys().collect();
+                let also_drained = possibly_kept
+                    .split_off(number_of_accounts_kept)
```

Comment:
Dumb question not looking at the unchanged code 🙃: is it possible for `possibly_kept` to be empty? If so, increasing `number_of_accounts_kept` above will lead to a panic here.

Reply:
nice catch, yes, it's possible